### DISCUSSION TITLE
Split code into client / server / common

### DISCUSSION THREAD

**Discussion by kberg** [bafolts/terraforming-mars#3939] [Ideas]
This is a look at what a running browser implementation has in terms of source code:

![image](https://user-images.githubusercontent.com/413481/112559016-6d228b00-8da6-11eb-93ea-30576f50bfeb.png)

Notice all those non-client directories it has? `deferredActions`? `inputs`? Those directories have nothing that the client should use, but they're included because some common components include them.

This impacts us in other ways: I was unable to move some code into `inputs` because they triggered dependencies on things like The Moon, and suddenly the client-side wouldn't compile because contents in The Moon which should never be compiled into the client were suddenly made available.

To separate them, we would need to make some architectural changes, keeping code that belongs on the server separate from what runs on the client, and creating a third separation just for data structures and code that's run in both places. This is not a simple process, but it will make a healthier product.

(It's possible this could be done more simply just by tightening down what goes into the build directory webpack uses.

**Comment by kberg** [comment]
I'm going to start experimenting with this by moving all code into three directories: `server/` `common/` and `client/`. with the only possible exception being the code that serves the entire app.

I'll also eventually implement some rules like,

1. Client code may import from `client/` and `common/` but not `server/`.
2. Server code mayt import from `server/` and `common/` but not `client/`.
3. Common code many not import from either `client` or `server`.

This will, for example, keep `deferredActions` away from the browser.

This will require some refactoring, but it will make for a better cleaner experience.

**Comment by kberg** [comment]
Here's an example of how easy it is to see how client and server are commingled. I moved some files that were most definitely server files in a `server` directory, and put some common material in a `common` directory.

Then I ran 

```
$ find src/components -type f | xargs egrep "import .*server"
src/components/CreateGameForm.ts:import {IGameData} from '../server/database/IDatabase';
```

`IDatabase` imports `Game` and now you're off to the races. Everything is a dependency on the client.

**Comment by kberg** [comment]
So I moved `IGameData` to `common` and cleaned up some types, and `IGameData.ts` looked like

```
import {GameId} from '../Game';

export interface IGameData {
    gameId: GameId;
    playerCount: number;
}
```

And of course, `Game` is server-side, so I created another file, `common/Types.ts` with this content

`export type GameId = string`

And now IGameData no longer relies on Game, and CreateGameForm no longer relies on a database interface.

**Comment by bafolts** [comment]
I think we are lucky and webpack is removing the unused classes for us during the build process. We may not be getting lucky in all situations though. If you search for `Database` in `build/main.js` you wont find any results. The `Game` class also doesn't end up in the build.

This will help with contributing and make it clearer which code is running where. Every `Card` may need to be split. We send all of the server logic for a card to the client. For example if you look in `build/main.js` you will see the `Ant` card includes functions like `canAct` which is only intended for server side usage.

In `build/main.js` you will see for `Ants` this server side logic:

```
    Ants.prototype.getVictoryPoints = function () {
        return Math.floor(this.resourceCount / 2);
    };
    Ants.prototype.play = function () {
        return undefined;
    };
    Ants.prototype.canAct = function (player) {
        if (player.game.isSoloMode())
            return true;
        return RemoveResourcesFromCard_1.RemoveResourcesFromCard.getAvailableTargetCards(player, this.resourceType).length > 0;
    };
```

**Comment by kberg** [comment]
I'm glad to read that about Webpack. And I'm happy to slowly do the work to get code we don't want out of there. And it might not even need the hard split like I'm suggesting (though it honestly wouldn't hurt for clarity.)

But how then could we prevent things creeping back in? Does webpack have an ability to raise an error if one of the inbound dependencies is annotated? Say, with an annotation we define that indicates something should only be on the serverr?`@server` ?

**Comment by kberg** [comment]
This is my new major effort. Many, many of the PRs will involve moving files from their current location to a directory called `src/common` and should be easy to review. Many more PRs will represent moving server-side code out of the client, but I'll have to write reasonable replacements (without duplicating code as much as possible.)

**Comment by kberg** [comment]
Once #4117 goes in, `src/common/` will have over 70 source files in it, none of it references code outside of `common`. With one or two exceptions, that's all the easy ones done. Next come the hard ones, and there's about 40 of them.

I won't list them here, but you can see them by running  `find src -name "*.vue" | xargs grep import | sed 's/.* from //' | sort | uniq | grep -v common | grep -v @/client | grep -v \.vue`

There's a couple of simple leftovers (like `constants`, `Timer`, `HowToPay`) but it's time for the real work to start.

**Comment by kberg** [comment]
This is what's left:

```
'@/cards/corporation/ICorporationCard';
'@/cards/ICard';
'@/cards/prelude/PreludeCard';
'@/genfiles/settings.json';
'@/turmoil/globalEvents/GlobalEventDealer';
'@/turmoil/globalEvents/IGlobalEvent';
```

These are the hardest ones. I know how to do it, but gah, it will still be hard work.

**Comment by kberg** [comment]
This is now done!
